### ISSUE TITLE
OAuth support for IMAP authentication

### ISSUE THREAD

**Issue by chibenwa** [linagora/james-project#4436]
## Why? 

OAuth is an industry standard and IMAP OAuth support is a recuring question we get in the community.

This is specified in https://datatracker.ietf.org/doc/html/rfc7628 .

## How? 

We will expose an option (disabled by default) to enable OAuth authentication in IMAP.

The IMAP configuration should:
  - URL of the OIDC supplier
  - Name of the claim acting as the login (email address)
  - JWKS URL of the OIDC provider

Also, the authentication should allow turning on / turning off plain authentication (in order to "force" using the OIDC server.)

Exemple:

```
    <imapserver enabled="true">
        <jmxName>imapserver-ssl</jmxName>
        <bind>0.0.0.0:993</bind>
        <connectionBacklog>200</connectionBacklog>
        <tls socketTLS="true" startTLS="false">
            <privateKey>file://conf/private.key</privateKey>
            <certificates>file://conf/certs.self-signed.csr</certificates>
        </tls>
        <connectionLimit>0</connectionLimit>
        <connectionLimitPerIP>0</connectionLimitPerIP>
        <idleTimeInterval>120</idleTimeInterval>
        <idleTimeIntervalUnit>SECONDS</idleTimeIntervalUnit>
        <enableIdle>true</enableIdle>
        <authentication>
            <plainAuthEnabled>true|false</plainAuthEnabled> <!-- defaults to true -->
            <oauth> <!-- ommiting this block would disable oauth. All fields are compulsory -->
                <jwksURL>https://example.com/jwks</jwksURL> 
                <oidcConfigurationURL>https://example.com/.well-known/openid-config</oidcConfigurationURL>
                <claim>mailAddress</claim>
            </oauth>
        </authentication>
    </imapserver>
```

of course the authentication part of the configuration file should be backed by a configuration POJO described in #xxxx .

Implement a OAUTH mechanism in `AuthenticateProcessor`, using the right `JWTVerifier` as created from the configuration. https://github.com/auth0/jwks-rsa-java can be used to get easily the public key out of a JWK endpoint.

## Examples from the spec

TLDR: a first roundtrip is needed for the client to discover id provider location:

```
   This IMAP example shows a failed exchange because of the empty
   Authorization header, which is how a client can query for the needed
   scope.  Note that line breaks are inserted for readability.

   S: * OK IMAP4rev1 Server Ready
   C: t0 CAPABILITY
   S: * CAPABILITY IMAP4rev1 AUTH=OAUTHBEARER SASL-IR
   S: t0 OK Completed
   C: t1 AUTHENTICATE OAUTHBEARER bixhPXVzZXJAZXhhbXBsZS5jb20sAW
         hvc3Q9c2VydmVyLmV4YW1wbGUuY29tAXBvcnQ9MTQzAWF1dGg9AQE=
   S: + eyJzdGF0dXMiOiJpbnZhbGlkX3Rva2VuIiwic2NvcGUiOiJleGFtcGxl
        X3Njb3BlIiwib3BlbmlkLWNvbmZpZ3VyYXRpb24iOiJodHRwczovL2V4
        YW1wbGUuY29tLy53ZWxsLWtub3duL29wZW5pZC1jb25maWd1cmF0aW9u
        In0=
   C: AQ==
   S: t1 NO SASL authentication failed

   The decoded initial client response is:

   n,a=user@example.com,^Ahost=server.example.com^A
   port=143^Aauth=^A^A

   The decoded server error response is:

  {
  "status":"invalid_token",
  "scope":"example_scope",
  "openid-configuration":"https://example.com/.well-known/openid-config"
  }

   The client responds with the required dummy response; "AQ==" is the
   base64 encoding of the ASCII value 0x01.  The same exchange using the
   IMAP-specific method of canceling an AUTHENTICATE command sends "*"
   and is shown below.

   S: * OK IMAP4rev1 Server Ready
   C: t0 CAPABILITY
   S: * CAPABILITY IMAP4rev1 AUTH=OAUTHBEARER SASL-IR IMAP4rev1
   S: t0 OK Completed
   C: t1 AUTHENTICATE OAUTHBEARER bixhPXVzZXJAZXhhbXBsZS5jb20sAW
        hvc3Q9c2VydmVyLmV4YW1wbGUuY29tAXBvcnQ9MTQzAWF1dGg9AQE=
   S: + eyJzdGF0dXMiOiJpbnZhbGlkX3Rva2VuIiwic2NvcGUiOiJleGFtcGxl
        X3Njb3BlIiwib3BlbmlkLWNvbmZpZ3VyYXRpb24iOiJodHRwczovL2V4
        YW1wbGUuY29tLy53ZWxsLWtub3duL29wZW5pZC1jb25maWd1cmF0aW9u
        In0=
   C: *
   S: t1 NO SASL authentication failed
```

Having discovered the location of the OpenId server the client can then get its token there and come back:

```
   This example shows a successful OAuth 2.0 bearer token exchange in
   IMAP.  Note that line breaks are inserted for readability.

   [Initial connection and TLS establishment...]
   S: * OK IMAP4rev1 Server Ready
   C: t0 CAPABILITY
   S: * CAPABILITY IMAP4rev1 AUTH=OAUTHBEARER SASL-IR
   S: t0 OK Completed
   C: t1 AUTHENTICATE OAUTHBEARER bixhPXVzZXJAZXhhbXBsZS5jb20sAWhv
         c3Q9c2VydmVyLmV4YW1wbGUuY29tAXBvcnQ9MTQzAWF1dGg9QmVhcmVyI
         HZGOWRmdDRxbVRjMk52YjNSbGNrQmhiSFJoZG1semRHRXVZMjl0Q2c9PQ
         EB
   S: t1 OK SASL authentication succeeded

   As required by IMAP [RFC3501], the payloads are base64 encoded.  The
   decoded initial client response (with %x01 represented as ^A and long
   lines wrapped for readability) is:

   n,a=user@example.com,^Ahost=server.example.com^Aport=143^A
   auth=Bearer vF9dft4qmTc2Nvb3RlckBhbHRhdmlzdGEuY29tCg==^A^A
```

## Definition of done

 - **MPT tests** that exercise oauth SASL mechanism

JWKS endpoint can be mocked using Mock HTTP server.

## Limitations

 - No revocation of token is supported so far (valid signed tokens can always be used)
 - Revoking a token to not end the IMAP sessions using that token.

**Comment by quantranhong1999** [comment]
> No revocation of token is supported so far (valid signed tokens can always be used)

Do we have a plan to support this in the future? 

> Revoking a token to not end the IMAP sessions using that token

Can you explain more about this? Is that even we somehow can revoke the token, we still don't have yet a mechanism to end the IMAP session using that invalidated token?

**Comment by chibenwa** [comment]
> Do we have a plan to support this in the future?

I'm npot sure this makes much sens for IMAP/SMTP

> Can you explain more about this? Is that even we somehow can revoke the token, we still don't have yet a mechanism to end the IMAP session using that invalidated token?

That's it!

**Comment by Arsnael** [comment]
it's not critical as well IMO. The JWT tokens issued by Keycloak for example are supposed to be short lived
